homeassistant/components/logbook/websocket_api.py

```diff
@@ -0,0 +1,263 @@
+"""Event parser and human readable log generator."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+from datetime import datetime as dt
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.components.recorder import get_instance
+from homeassistant.components.websocket_api import messages
+from homeassistant.components.websocket_api.connection import ActiveConnection
+from homeassistant.components.websocket_api.const import JSON_DUMP
+from homeassistant.const import EVENT_STATE_CHANGED
+from homeassistant.core import CALLBACK_TYPE, Event, HomeAssistant, callback
+from homeassistant.helpers.event import async_track_state_change_event
+import homeassistant.util.dt as dt_util
+
+from .helpers import async_determine_event_types
+from .models import async_event_to_row
+from .processor import EventProcessor
+
+MAX_PENDING_LOGBOOK_EVENTS = 2048
+EVENT_COALESCE_TIME = 1
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_setup(hass: HomeAssistant) -> None:
+    """Set up the logbook websocket API."""
+    websocket_api.async_register_command(hass, ws_get_events)
+    websocket_api.async_register_command(hass, ws_event_stream)
+
+
+def _ws_formatted_get_events(
+    msg_id: int,
+    start_day: dt,
+    end_day: dt,
+    formatter: Callable[[int, Any], dict[str, Any]],
+    event_processor: EventProcessor,
+) -> tuple[str, dt | None]:
+    """Fetch events and convert them to json in the executor."""
+    events = event_processor.get_events(start_day, end_day)
+    last_time = None
+    if events:
+        last_time = dt_util.utc_from_timestamp(events[-1]["when"])
+    result = formatter(msg_id, events)
+    return JSON_DUMP(result), last_time
+
+
+async def async_stream_events(
+    hass: HomeAssistant,
+    connection: ActiveConnection,
+    msg_id: int,
+    stream_queue: asyncio.Queue[Event],
+    event_processor: EventProcessor,
+    last_time_from_db: dt,
+) -> None:
+    """Stream events from the queue."""
+    instance = get_instance(hass)
+    # Fetch any events from the database that have
+    # not been committed since the original fetch
+    await asyncio.sleep(instance.commit_interval)
+    message, last_time = await instance.async_add_executor_job(
+        _ws_formatted_get_events,
+        msg_id,
+        last_time_from_db,
+        dt_util.utcnow(),
+        messages.event_message,
+        event_processor,
+    )
+    if last_time:
+        connection.send_message(message)
+    else:
+        last_time = last_time_from_db
```

Comment:
I think we can delete all of this now as the event listeners are setup before the call now so we will already have the data waiting on the queue